# **Лекция 6. Методы агрегирования при анализе СМО**

 - Пример задачи и ее точное решение
 - Метод агрегирования с использованием параметров связи, оценка погрешности
 - Метод укрупнения состояний, оценка погрешности

## **$\S 1.$ Пример задачи и ее точное решение**

В лекции 5 мы видели, что даже в простейших системах массового обслуживания при большом количестве параметров получить точное решение становится довольно сложно. Если же говорить о реальных моделях СМО, то это становится просто нереально. Нужны какие-то рациональные способы упрощения модели с минимальными потерями точности.

Рассмотрим в качестве примера расширение задачи лабораторной работы: 

**$K$ программистов могут писать программы для выполнения на одном из $M$ серверов, при этом программа не попадает сразу на сервер, а обрабатывается на одном из $N$ специальных компьютеров, которые проверяют отсутствие вирусов. Интенсивность работы программистов $\lambda$, интенсивность работы компьютеров-антивирусов $\nu$, интенсивность работы основных серверов $\mu$, программа оказывается с вирусом с вероятностью $p$. Если программа с вирусом, она получает отказ обслуживания на основных серверах. Ни в каких частях системы очередь не образуется.** 

Будем рассматривать трехмерный вектор состояний системы: $\zeta(t)=(\zeta_1(t),\zeta_2(t), \zeta_3(t))$, где $\zeta_1(t)$ - количество программистов, которые пишут программу в момент времени $t$, $\zeta_2(t)$ - количество программ, которые проверяются на наличие вирусов, $\zeta_3(t)$ - количество программ, которые выполняются на основных серверах.

Рассмотрим только частный случай: $K=2, M=2, N=2$. Введем состояния:

$S_{200}$ - все программисты пишут программу, сервера и компьютеры свободны;

$S_{110}$ - один программист пишет, одна программа проверяется на вирус;

$S_{101}$ - один программист пишет, одна программа выполняется на сервере;

$S_{011}$ - оба программиста ожидают ответа, одна программа проверяется на вирус, другая выполняется на сервере;

$S_{020}$ - оба программиста ожидают ответа, обе программы проверяются на вирус;

$S_{002}$ - оба программиста ожидают ответа, обе программы выполняются на сервере.

Граф состояний:

<center><img src="ris_1.png" width=700 height=700></center>

По данному графу состояний можно составить систему дифференциальных уравнений, а также найти финальные вероятности. Однако при увеличении $K, M, N$ размерность задачи очень быстно возрастает. Уже при $K=N=M=10$ нарисовать граф состояний и следовательно составить уравнения для финальных вероятностей становится весьма сложно, а если добавить еще и возможность образования очереди, точное аналитическое решение становится практически недоступным.

В нашем случае попробуем вначале найти точное решение, составив уравнения Колмогорова:

$$\begin{cases}
  \frac{dP_{200}(t)}{dt}=-2\lambda\cdot P_{200}(t)+p\cdot \nu\cdot P_{110}(t)+\mu\cdot P_{101}(t)\\
  \frac{dP_{110}(t)}{dt}=-(\lambda+ \nu)\cdot P_{110}(t)+\mu\cdot P_{011}(t)+2p\nu\cdot P_{020}(t)+2\lambda\cdot P_{200}(t)\\
  \frac{dP_{020}(t)}{dt}=\lambda\cdot P_{110}(t)-2\nu\cdot P_{020}(t)\\
  \frac{dP_{101}(t)}{dt}=(1-p)\nu\cdot P_{110}(t)+p\nu\cdot P_{011}(t)+2\mu\cdot P_{002}(t)-(\lambda+\mu)\cdot P_{101}(t)\\
  \frac{dP_{011}(t)}{dt}=2(1-p)\nu\cdot P_{020}(t)+\lambda\cdot P_{101}(t)-(\mu+\nu)\cdot P_{011}(t)\\
  \frac{dP_{002}(t)}{dt}=-2\mu\cdot P_{002}(t)+(1-p)\nu\cdot P_{011}(t)\\
  P_{200}(t)+P_{110}(t)+P_{020}(t)+P_{101}(t)+P_{011}(t)+P_{002}(t)=1
\end{cases}$$

Зададимся некоторыми конкретными значениями $\lambda=0.8, \mu=1.2, \nu=1, p=0.2$ и найдем финальные вероятности, решив СЛАУ:

In [11]:
library(rim)

In [12]:
maxima.get("lambda: 0.8$")
maxima.get("mu: 1.2$") 
maxima.get("nu:1$")
maxima.get("p:0.2$")

In [13]:
maxima.options(engine.format = "latex", 
           engine.label = TRUE,
           inline.format = "latex", 
           inline.label = FALSE)

In [14]:
maxima.get("solve([-2*lambda*P_2_0_0+p*nu*P_1_1_0+mu*P_1_0_1=0,
       -(lambda+nu)*P_1_1_0+mu*P_0_1_1+2*p*nu*P_0_2_0+2*lambda*P_2_0_0=0,
       lambda*P_1_1_0-2*nu*P_0_2_0=0,
       (1-p)*nu*P_1_1_0+p*nu*P_0_1_1+2*mu*P_0_0_2-(lambda+mu)*P_1_0_1=0,
       -2*mu*P_0_0_2+(1-p)*nu*P_0_1_1=0,
       P_2_0_0+P_1_1_0+P_0_2_0+P_1_0_1+P_0_1_1+P_0_0_2=1],
       [P_2_0_0,P_1_1_0,P_0_2_0,P_1_0_1,P_0_1_1,P_0_0_2]);")

rat: replaced 1.2 by 6/5 = 1.2
rat: replaced 0.2 by 1/5 = 0.2
rat: replaced -1.6 by -8/5 = -1.6
rat: replaced 1.2 by 6/5 = 1.2
rat: replaced 0.4 by 2/5 = 0.4
rat: replaced -1.8 by -9/5 = -1.8
rat: replaced 1.6 by 8/5 = 1.6
rat: replaced 0.8 by 4/5 = 0.8
rat: replaced 2.4 by 12/5 = 2.4
rat: replaced 0.2 by 1/5 = 0.2
rat: replaced -2.0 by -2/1 = -2.0
rat: replaced 0.8 by 4/5 = 0.8
rat: replaced -2.4 by -12/5 = -2.4
rat: replaced 0.8 by 4/5 = 0.8



(%o11) [[P_2_0_0 = 9/49,P_1_1_0 = 72/245,P_0_2_0 = 144/1225,P_1_0_1 = 48/245,
         P_0_1_1 = 192/1225,P_0_0_2 = 64/1225]]

In [15]:
maxima.get("float(solve([-2*lambda*P_2_0_0+p*nu*P_1_1_0+mu*P_1_0_1=0,
       -(lambda+nu)*P_1_1_0+mu*P_0_1_1+2*p*nu*P_0_2_0+2*lambda*P_2_0_0=0,
       lambda*P_1_1_0-2*nu*P_0_2_0=0,
       (1-p)*nu*P_1_1_0+p*nu*P_0_1_1+2*mu*P_0_0_2-(lambda+mu)*P_1_0_1=0,
       -2*mu*P_0_0_2+(1-p)*nu*P_0_1_1=0,
       P_2_0_0+P_1_1_0+P_0_2_0+P_1_0_1+P_0_1_1+P_0_0_2=1],
       [P_2_0_0,P_1_1_0,P_0_2_0,P_1_0_1,P_0_1_1,P_0_0_2]));")

rat: replaced 1.2 by 6/5 = 1.2
rat: replaced 0.2 by 1/5 = 0.2
rat: replaced -1.6 by -8/5 = -1.6
rat: replaced 1.2 by 6/5 = 1.2
rat: replaced 0.4 by 2/5 = 0.4
rat: replaced -1.8 by -9/5 = -1.8
rat: replaced 1.6 by 8/5 = 1.6
rat: replaced 0.8 by 4/5 = 0.8
rat: replaced 2.4 by 12/5 = 2.4
rat: replaced 0.2 by 1/5 = 0.2
rat: replaced -2.0 by -2/1 = -2.0
rat: replaced 0.8 by 4/5 = 0.8
rat: replaced -2.4 by -12/5 = -2.4
rat: replaced 0.8 by 4/5 = 0.8



(%o12) [[P_2_0_0 = 0.1836734693877551,P_1_1_0 = 0.2938775510204082,
         P_0_2_0 = 0.11755102040816326,P_1_0_1 = 0.19591836734693877,
         P_0_1_1 = 0.156734693877551,P_0_0_2 = 0.052244897959183675]]

Найдем основные характеристики системы:
- **среднее число заявок в системе**:
$$L_{\text{сист}}=1\cdot (P_{110}+ P_{101})+2\cdot (P_{020}+P_{011}+P_{002})\approx 1.14285714285714$$

 - **абсолютная пропускная способность** (среднее число заявок, обслуженных в единицу времени): для определения абсолютной пропускной способности учтем, что заявки-программы, в которых обнаружен вирус не получают полного обслуживания (отклоняются), т.е. нас интересует среднее число заявок получивших полное обслуживание в единицу времени. Для замкнутых систем абсолютная пропускная способность выражается как
 $$\lambda'=P_{\text{зан}}\cdot \mu$$
 $P_{\text{зан}}$ - вероятность, что канал занят;
 $\mu$ - интенсивность обслуживания канала.
 
 В нашем случае на выходе два канала, но они работают не во всех состояниях, а только в $S_{101}$ с интенсивностью $\mu$, в состоянии $S_{011}$ тоже с интенсивностью $\mu$, и в состоянии $S_{002}$ с интенсивностью $2\mu$.
 $$\lambda'=\mu\cdot(P_{101}+P_{011})+2\mu\cdot P_{002}\approx 0.548571428571429$$

- *средняя интенсивность суммарного входного потока*:
 $$\Lambda=(k-L_{\text{сист}})\cdot \lambda=(2-1.142857142857143)\cdot 0.8\approx 0.685714285714286$$

- **среднее время пребывания заявки в системе**:
$$T_{\text{сист}}=\frac{L_{\text{сист}}}{\Lambda}\approx 1.666$$


Теперь продемонстрируем реализацию простой иммитационной модели и проведем моделирование.

In [16]:
lambda<-0.8; mu<-1.2; nu<-1; p<-0.2

In [18]:
proger<- rexp(2, lambda) #Время написания программ для двух программистов
virus<-c(0,0,0,0) #первый индекс - номер программиста, от которого проверка на вирус на первой ЭВМ, второй индекс - время обслуживания на 1 ЭВМ
                  #третий индекс -номер програмиста, от которого проверка на вирус на второй ЭВМ, четвертый индекс - время обслуживания на 2 ЭВМ
server<-c(0,0,0,0) #первый индекс - номер программиста, от которого выполнение на первом сервере, второй индекс - время обслуживания на 1 сервере
                  #третий индекс номер программиста, от которого выполнение на втором сервере, четвертый индекс - время обслуживания на 2 сервере
T<-10000 #Суммарное модельное время
h<-0.001 #Шаг приращения модельного времени
t<-0 # Текущее время

obs<-0 #Количество заявок, получивших полное обслуживание
request_col<-0 #Суммарное количество заявок в системе в каждый квант времени
tsum<-0 # Суммарное время пребывания всех заявок в системе
requestsum<-0 # Считаем суммарное количество заявок, поступивших в СМО

In [19]:
while(t<T){
  if ((proger[1]<= t) & (proger[1] !=0))
      {
      requestsum<-requestsum+1
      if (virus[1] == 0)
          {virus[1]<-1
           virus[2]<-t+rexp(1,nu)
           proger[1]<- 0
          }
      else
          {if (virus[3] == 0)
          {virus[3]<-1
           virus[4]<-t+rexp(1,nu)
           proger[1]<- 0
          }}  
      }
  if ((proger[2]<= t) & (proger[2] !=0))
      {
      requestsum<-requestsum+1
      if (virus[1] == 0)
          {virus[1]<-2
           virus[2]<-t+rexp(1,nu)
           proger[2]<- 0
          }
      else
          {if (virus[3] == 0)
          {virus[3]<-2
           virus[4]<-t+rexp(1,nu)
           proger[2]<- 0
          }}  
      }
  if ((virus[2]<=t) &(virus[2] !=0)){
      u<-runif(1)
      if (u > p){
          if (server[1] == 0){
                server[1]<-virus[1]
                server[2]<-t+rexp(1,mu)
                virus[1]<-0;virus[2]<-0;
              }
          else if (server[3] == 0){
                server[3]<-virus[1]
                server[4]<-t+rexp(1,mu)
                virus[1]<-0;virus[2]<-0;
              }
          
          }
      else {
          proger[virus[1]]<-t +rexp(1,lambda)
          virus[1]<-0;virus[2]<-0;
       }
  }
  if ((virus[4]<=t) & (virus[4]!=0)){
      u<-runif(1)
      if (u > p){
          if (server[1] == 0){
                server[1]<-virus[3] 
                server[2]<-t+rexp(1,mu)
                virus[3]<-0;virus[4]<-0;
              }
          else if (server[3] == 0){
                server[3]<-virus[3]
                server[4]<-t+rexp(1,mu)
                virus[3]<-0;virus[4]<-0;
              }
          
          }
      else {
          proger[virus[3]]<-t +rexp(1,lambda)
          virus[3]<-0;virus[4]<-0;
       }
  }
  if ((server[2]<=t) & (server[2] !=0)){
      proger[server[1]]<-t +rexp(1,lambda)
      server[1]<-0;server[2]<-0;
      obs<-obs+1
  }
  if ((server[4]<=t) & (server[4]!=0)){
      proger[server[3]]<-t +rexp(1,lambda)
      server[3]<-0;server[4]<-0;
      obs<-obs+1
  }
  request_col<-request_col+length(proger[proger==0])
  tsum<-tsum+length(proger[proger==0])*h
  t<-t+h
}

In [25]:
cat("Среднее число заявок в системе:",request_col/(T/h),"\n")
cat("Средняя интенсивность суммарного входного потока:",requestsum/T,"\n")
cat("Абсолютная пропускная способность:",obs/(T),"\n")
cat("Среднее время пребывания заявки в системе:",tsum/requestsum)

Среднее число заявок в системе: 1.147686 
Средняя интенсивность суммарного входного потока: 0.6917 
Абсолютная пропускная способность: 0.5539 
Среднее время пребывания заявки в системе: 1.659225

Таким образом, можно констатировать, что результаты совпали.

## $\S 2.$ **Метод агрегирования с использованием параметров связи, оценка погрешности**

Рассмотрим теперь упрощенные подходы к анализу нашей системы. Первым из таких подходов рассмотрим метод агрегирования на основе параметров связи.

Суть метода состоит в том, чтобы анализ сложной системы производить по частям, с помощью укрупненных (агрегированных) моделей. В каждой такой модели подробно представлена только некоторая часть системы, а влияние остальных частей отражается **обобщенным параметром (параметром связи)**. Для этого используется **принцип эквивалентности потоков** - часть системы заменяется агрегированным узлом, в котором интенсивность обслуживания (параметр связи) зависит от
числа заявок в узле. Принцип эквивалентности потоков заключается в том, что агрегированный узел при любом числе заявок в нем должен обеспечивать такой же поток во внешнюю сеть (такую же пропускную способность), что и заменяемая им подсистема.

Предложим декомпозицию нашей задачи в виде двух моделей.

В первой модели совокупность серверов и компьютеров по проверке наличия вирусов заменим интенсивностями обслуживания $\mu(i)$. Тогда модель сводится к обычной схеме рождения-гибели:

<center><img src="ris_2.png" width=700 height = 700></center>

Основные параметры системы могут быть определены из соотношений:

$$P_0=\left(1+\frac{K\lambda}{\mu(1)}+\frac{K(K-1)\lambda^2}{\mu(1)\mu(2)}+\ldots+\frac{K!\lambda^K}{\prod_{k=1}^K\mu(k)}\right)^{-1}\tag{2.1}$$
$$P_1=P_0\frac{K\lambda}{\mu(1)}\tag{2.2}$$
$$P_2=P_0\frac{K(K-1)\lambda^2}{\mu(1)\mu(2)}\tag{2.3}$$
$$\ldots$$
$$L_{\text{сист}}=\sum_{k=1}^K k\cdot P_k\tag{2.4}$$
$$T_{\text{сист}}=\frac{L_{\text{сист}}}{\lambda(K-L_{\text{сист}})}\tag{2.5}$$
$$\mu_{\text{ср}}=\sum_{k=1}^K\mu(k)\cdot P_k\tag{2.6}$$

Для нахождения параметров связи $\mu(i)$ рассмотрим вторую модель - за состояния системы примем количество заявок, находящихся на основных серверах. Предполагаем, что в системе циркулирует $L_{\text{сист}}$ заявок, проводим расчеты при всех его возможных значениях $L_{\text{сист}}=1,2,3,\dots,K$. Имеем следующий граф состояний для фазы обслуживания:

<center><img src="ris_3.png" width=700 height=700></center>

$$\nu_n=(1-p)\nu\cdot \min{(N,u-n)}, n=0,\ldots,u-1\tag{2.7}$$
$$\mu_n=\mu\cdot \min{(M, n)}, n=1,2, \ldots,u\tag{2.8}$$

$$\pi_0=\left(1+\frac{\nu_0}{\mu_1}+\frac{\nu_0\nu_1}{\mu_1\mu_2}+\ldots\right)^{-1}\tag{2.9}$$
$$\pi_1=\pi_0\cdot \frac{\nu_0}{\mu_1}\tag{2.10}$$
$$\pi_2=\pi_0\cdot \frac{\nu_0\nu_1}{\mu_1\mu_2}\tag{2.11}$$
$$\ldots$$
$$\mu(u)=\sum_{n=1}^{u}\mu_n\pi_n\tag{2.12}$$

Расчеты по формулам (2.7-2.12) повторяют для каждого $u=\overline{1,K}$. Полученные значения $\mu(u)$ используют в формулах (2.1-2.6).

Рассмотрим расчеты по данным формулам на примере нашей задачи: $$M=N=K=2, \lambda=0.8, \mu=1.2, \nu=1, p=0.2$$

Последовательность расчета будет следующей: 1. по формулам (2.7) - (2.12) вычисляем параметры связи $\mu(u)$; 2. по формулам (2.1) - (2.6) вычисляем основные показатели качества СМО.

(2.7): $$\nu_n=(1-p)\nu\cdot \min{(N,u-n)}, n=0,\ldots,u-1$$
(2.8): $$\mu_n=\mu\cdot \min{(M, n)}, n=1,2, \ldots,u$$

$$\fbox{u=1}$$
$$\nu_0=0.8\cdot 1\cdot \min(2,1-0)=0.8$$
$$\mu_1=1.2\cdot \min(2,1)=1.2$$
$$\pi_0=\left(1+\frac{\nu_0}{\mu_1}\right)^{-1}=\left(1+\frac{0.8}{1.2}\right)^{-1}\approx 0.6$$
$$\pi_1=\pi_0\cdot \frac{\nu_0}{\mu_1}=0.6\cdot \frac{0.8}{1.2}\approx 0.4$$
$$\mu(1)=\mu_1\cdot \pi_1=1.2\cdot 0.4\approx 0.48$$

$$\fbox{u=2}$$
$$\nu_0=0.8\cdot 1\cdot \min(2,2-0)=0.8\cdot 2=1.6$$
$$\mu_1=1.2\cdot \min(2,1)=1.2$$
$$\nu_1 = 0.8\cdot 1\cdot \min(2,2-1)=0.8$$
$$\mu_2=1.2\cdot \min(2,2)=2.4$$
$$\pi_0=\left(1+\frac{\nu_0}{\mu_1}+\frac{\nu_0\nu_1}{\mu_1\mu_2}\right)^{-1}\approx 0.36$$
$$\pi_1=\pi_0\cdot \frac{\nu_0}{\mu_1}\approx 0.48$$
$$\pi_2=\pi_0\cdot \frac{\nu_0\nu_1}{\mu_1\mu_2}\approx 0.16$$
$$\mu(2)=\pi_1\cdot \mu_1+\pi_2\cdot\mu_2=1.2\cdot 0.48+2.4\cdot 0.16=0.96$$

Теперь перейдем к первой модели и выполним расчеты показателей качества СМО:

$$P_0=\left(1+\frac{K\lambda}{\mu(1)}+\frac{K(K-1)\lambda^2}{\mu(1)\mu(2)}\right)^{-1}=\left(1+\frac{2\cdot 0.8}{0.48}+\frac{2\cdot(2-1)\cdot 0.8^2}{0.48\cdot 0.96}\right)^{-1}\approx 0.14$$
$$P_1=P_0\frac{K\lambda}{\mu(1)}=0.14\cdot \frac{2\cdot 0.8}{0.48}\approx 0.47$$
$$P_2=P_0\frac{K(K-1)\lambda^2}{\mu(1)\mu(2)}=0.14\cdot \frac{2\cdot 1\cdot 0.8^2}{0.48\cdot 0.96}\approx 0.39$$
$$L_{\text{сист}}=\sum_{k=1}^K k\cdot P_k=1\cdot 0.47+2\cdot 0.39\approx 1.25$$
$$T_{\text{сист}}=\frac{L_{\text{сист}}}{\lambda(K-L_{\text{сист}})}=\frac{1.25}{0.8(2-1.25)}\approx 2.08$$
$$\mu_{\text{ср}}=\sum_{k=1}^K\mu(k)\cdot P_k=0.48\cdot 0.47+0.96\cdot 0.39\approx 0.6$$

Ниже приведено сопоставление точных значений и значений, полученный по упрощенной модели:

|$L_{\text{сист}}$|$L_{\text{сист}}^{\text{поток}}$|$\lambda'$|$\lambda'^{\text{поток}}$|$T_{\text{сист}}$|$T_{\text{сист}}^{\text{поток}}$|
|--|--|--|--|--|--|
|1.14|1.25|0.55|0.6|1.67|2.08|

Как видно из полученных значений, с помощью метода на основе эквивалентности потоков достигается приемлемая для инженерных расчетов точность.

## **$\S 3.$ Метод укрупнения состояний, оценка погрешности**

Еще одним методом агрегирования является метод укрупнения состояний. **Суть метода состоит в том, что на графе состояний несколько состояний объединяют в одно.** Иногда такое укрупнения оказывается **полностью эквивалентным исходному графу**: система уравнений относительно вероятностей состояний допускает сложение некоторой группы уравнений с вынесением за скобки общих множителей (интенсивностей перехода). В результате в скобках получается сумма вероятностей некоторой группы состояний. Эта группа и объявляется вероятностью макросостояния (укрупненного состояния). Часто прибегают к неполному эквивалентированию, тогда речь идет о **квазиэквивалентном укрупнении**. Важно отметить, что этот подход касается именно устоявшихся, стационарных состояний.

Рассмотрим в качестве примера граф, допускающий эквивалентное укрупнение:

**Одноядерный микропроцессор ставит выполнямые задачи в очередь, иногда микропроцессор переключается на выполнение своих задач.  Задача, которая выполняется в этот момент, помещается в очередь. Всего в систему может поступить не более трех задач. Интенсивность потока задач $\lambda$, потока выполнения задач $\mu$, интенсивность перечлючения операционной системы на выполнение своих задач $\nu$, интенсивность выполнения своих задач $\gamma$. Определить возможные состояния системы и финальные вероятности нахождения в них.**

Граф состояний:

$S_0$ - нет выполняемых задач, микропроцессор готов выполнять внешние задачи;

$S_1$ - микропроцессор занят выполнением внешней задачи, очереди нет;

$S_2$ - микропроцессор занят выполнением внешней задачи, в очереди одна задача;

$S_3$ - микропроцессор занят выполнением внешней задачи, в очереди две задачи;

$S_4$ - микропроцессор переключился на свои задачи, в очереди нет задач;

$S_5$ - микропроцессор переключился на свои задачи, в очереди одна задача;

$S_6$ - микропроцессор переключился на свои задачи, в очереди две задачи.

$S_7$ - микропроцессор переключился на свои задачи, в очереди три задачи.

<center><img src="ris_4.png" width=700 height=700></center>

Составим систему уравнений для стационарного режима: 

$$S_0:-(\lambda+\nu)P_0+\mu\cdot P_1+\gamma\cdot P_4=0\tag{3.1}$$
$$S_1:\lambda\cdot P_0-(\lambda+\mu+\nu)\cdot P_1+\mu\cdot P_2+\gamma\cdot P_5=0\tag{3.2}$$
$$S_2:\lambda\cdot P_1-(\lambda+\mu+\nu)\cdot P_2+\mu\cdot P_3+\gamma\cdot P_6=0\tag{3.3}$$
$$S_3:\lambda\cdot P_2-(\mu+\nu)\cdot P_3+\gamma\cdot P_7=0\tag{3.4}$$
$$S_4:\nu\cdot P_0-(\lambda+\gamma)\cdot P_4=0\tag{3.5}$$
$$S_5:\nu\cdot P_1+\lambda\cdot P_4-(\lambda+\gamma)\cdot P_5=0\tag{3.6}$$
$$S_6:\nu\cdot P_2+\lambda\cdot P_5-(\gamma+\lambda)\cdot P_6=0\tag{3.7}$$
$$S_7:\nu\cdot P_3+\lambda\cdot P_6-\gamma\cdot P_7=0\tag{3.8}$$

Попробуем выполнить так называемое **горизонтальное укрупнение**: объединим в одно состояние $W_0=\{S_0,S_1, S_2, S_3\}$ - сервер занят обслуживанием, $W_1=\{S_4,S_5,S_6,S_7\}$ - сервер занят выполнением своих задач.

Сгруппируем коэффициенты уравнения в таблицу так, чтобы объединенные состояния стояли рядом:



<table border=1>
    <tr>
        <td></td>
        <td>$S_0$</td>
        <td>$S_1$</td>
        <td>$S_2$</td>
        <td>$S_3$</td>
        <td>$S_4$</td>
        <td>$S_5$</td>
        <td>$S_6$</td>
        <td>$S_7$</td>
    </tr>
    <tr>
        <td>$S_0$</td>
        <td>$-(\lambda+\nu)$</td>
        <td>$\mu$</td>
        <td>0</td>
        <td>0</td>
        <td>$\gamma$</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
    </tr>
    <tr>
        <td>$S_1$</td>
        <td>$\lambda$</td>
        <td>$-(\lambda+\nu+\mu)$</td>
        <td>$\mu$</td>
        <td>0</td>
        <td>0</td>
        <td>$\gamma$</td>
        <td>0</td>
        <td>0</td>
    </tr>
    <tr>
        <td>$S_2$</td>
        <td>0</td>
        <td>$\lambda$</td>
        <td>$-(\lambda+\nu+\mu)$</td>
        <td>$\mu$</td>
        <td>0</td>
        <td>0</td>
        <td>$\gamma$</td>
        <td>0</td>
    </tr>
    <tr>
        <td>$S_3$</td>
        <td>0</td>
        <td>0</td>
        <td>$\lambda$</td>
        <td>$-(\mu+\nu)$</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
        <td>$\gamma$</td>
    </tr>
    <tr>
        <td>$S_4$</td>
        <td>$\nu$</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
        <td>$-(\lambda+\gamma)$</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
    </tr>
    <tr>
        <td>$S_5$</td>
        <td>0</td>
        <td>$\nu$</td>
        <td>0</td>
        <td>0</td>
        <td>$\lambda$</td>
        <td>$-(\lambda+\gamma)$</td>
        <td>0</td>
        <td>0</td>
    </tr>
    <tr>
        <td>$S_6$</td>
        <td>0</td>
        <td>0</td>
        <td>$\nu$</td>
        <td>0</td>
        <td>0</td>
        <td>$\lambda$</td>
        <td>$-(\lambda+\gamma)$</td>
        <td>0</td>
    </tr>
    <tr>
        <td>$S_7$</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
        <td>$\nu$</td>
        <td>0</td>
        <td>0</td>
        <td>$\lambda$</td>
        <td>$-\gamma$</td>
    </tr>
</table>

Из таблицы легко увидеть, что переход из множества состояний $W_0$ в множество состояний $W_1$ происходит с интенсивностью $\gamma$, а обратно с интенсивностью $\nu$.

К этому можно придти и выполняя преобразования уравнений.

Сложим правые и левые части уравнений (3.1), (3.2), (3.3), (3.4):
$$-\nu(P_0+P_1+P_2+P_3)+\gamma(P_4+P_5+P_6+P_7)=0$$

In [47]:
maxima.get("kill(lambda,nu,mu,gamma)$")
maxima.get("expand(-(lambda+nu)*P0+mu*P1+gamma*P4+
      lambda*P0-(lambda+mu+nu)*P1+mu*P2+gamma*P5+
      lambda*P1-(lambda+mu+nu)*P2+mu*P3+gamma*P6+
      lambda*P2-(mu+nu)*P3+gamma*P7);")

(%o48) P7*gamma+P6*gamma+P5*gamma+P4*gamma-P3*nu-P2*nu-P1*nu-P0*nu

Или сложим уравнения (3.5), (3.6), (3.7), (3.8):

In [48]:
maxima.get("kill(lambda,mu,nu,gamma)$")
maxima.get("expand(nu*P0-(lambda+gamma)*P4+
      nu*P1+lambda*P4-(lambda+gamma)*P5+
      nu*P2+lambda*P5-(gamma+lambda)*P6+
      nu*P3+lambda*P6-gamma*P7);")

(%o50) -(P7*gamma)-P6*gamma-P5*gamma-P4*gamma+P3*nu+P2*nu+P1*nu+P0*nu

Фактически получили тоже самое уравнение:
$$\nu(P_0+P_1+P_2+P_3)-\gamma(P_4+P_5+P_6+P_7)=0$$

Таким образом, от исходной системы можно перейти к упрощенному (укрупненном) графу состояний:

<center><img src="ris_6.png" width=200 height=300></center>

$$\gamma\cdot P_1^h-\nu\cdot P_0^h=0$$
$$P_0^h+P_1^h=1$$
$$P_0^h=P_0+P_1+P_2+P_3$$
$$P_1^h=P_4+P_5+P_6+P_7$$
Убеждаемся в полной эквивалентности укрупненных состояний:

In [49]:
maxima.get("lambda:1$")
maxima.get("nu:0.5$")
maxima.get("mu:0.8$")
maxima.get("gamma:1.2$")
maxima.get("solve([gamma*P1h-nu*P0h=0,P0h+P1h=1],[P0h,P1h]);")

rat: replaced -0.5 by -1/2 = -0.5
rat: replaced 1.2 by 6/5 = 1.2



(%o55) [[P0h = 12/17,P1h = 5/17]]

Таким образом, получилось:
$$P_0^h=\frac{12}{17},P_1^h=\frac{5}{17}$$

Найдем точное решение для системы уравнений (3.1)-(3.8):

In [50]:
maxima.get("solve([-(lambda+nu)*P0+mu*P1+gamma*P4=0,
      lambda*P0-(lambda+mu+nu)*P1+mu*P2+gamma*P5=0,
      lambda*P1-(lambda+mu+nu)*P2+mu*P3+gamma*P6=0,
      lambda*P2-(mu+nu)*P3+gamma*P7=0,
      nu*P0-(lambda+gamma)*P4=0,
      nu*P1+lambda*P4-(lambda+gamma)*P5=0,
      nu*P2+lambda*P5-(gamma+lambda)*P6=0,
      P0+P1+P2+P3+P4+P5+P6+P7=1],
      [P0,P1,P2,P3,P4,P5,P6,P7]);")

rat: replaced -1.5 by -3/2 = -1.5
rat: replaced 0.8 by 4/5 = 0.8
rat: replaced 1.2 by 6/5 = 1.2
rat: replaced -2.3 by -23/10 = -2.3
rat: replaced 0.8 by 4/5 = 0.8
rat: replaced 1.2 by 6/5 = 1.2
rat: replaced -2.3 by -23/10 = -2.3
rat: replaced 0.8 by 4/5 = 0.8
rat: replaced 1.2 by 6/5 = 1.2
rat: replaced -1.3 by -13/10 = -1.3
rat: replaced 1.2 by 6/5 = 1.2
rat: replaced 0.5 by 1/2 = 0.5
rat: replaced -2.2 by -11/5 = -2.2
rat: replaced 0.5 by 1/2 = 0.5
rat: replaced -2.2 by -11/5 = -2.2
rat: replaced 0.5 by 1/2 = 0.5
rat: replaced -2.2 by -11/5 = -2.2



(%o56) [[P0 = 2725888/35071493,P1 = 4181760/35071493,P2 = 6767200/35071493,
         P3 = 11081500/35071493,P4 = 619520/35071493,P5 = 1232000/35071493,
         P6 = 2098000/35071493,P7 = 6365625/35071493]]

In [51]:
maxima.get("2725888/35071493+4181760/35071493+6767200/35071493+11081500/35071493;")

(%o57) 12/17

In [52]:
maxima.get("619520/35071493+1232000/35071493+2098000/35071493+6365625/35071493;")

(%o58) 5/17

Таким образом, у нас получилось **полностью эквивалентное преобразование**. В результате решения можно оценить какую долю времени сервер занят выполнением внешних и внутренних задач, не прибегая к решению системы (3.1)-(3.8). Из составленной таблицы легко установить **критерии полной эквивалентности в методе укрупнения состояний**: переходы из одного укрупненного множества состояний в другое укрупненное множество состояний должны происходить с одиноковой интенсивностью. 

В нашем случае из любого состояния $S_0, S_1, S_2, S_3$ мы переходим в любое состояние множества $S_4, S_5, S_6, S_7$ с интенсивностью $\nu$ и обратно с интенсивностью $\gamma$.

Однако на практике добиться полностью эквивалентных состояний удается редко. Поэтому часто речь ведут про **квазиэквивалентное укрупнение состояний**.

В качестве примера квазиэквивалентного укрупнения состояний выполним на нашем графе так называемое вертикальное укрупнение состояний. Оно позволит оценить основные характеристики системы:

$$P_0^v=P_0+P_4, P_1^v=P_1+P_5, P_2^v=P_2+P_6, P_3^v=P_3+P_7$$

<center><img src="ris_7.png" width=700 height=700></center>

Т.е. будем говорить про четыре множества укрупненных состояний: $W_0=\{S_0, S_4\}, W_1=\{S_1, S_5\}, W_2=\{S_2, S_6\}, W_3=\{S_3, S_7\}$. Перегруппируем соответствующим образом нашу таблицу:

<table border=1>
    <tr>
        <td></td>
        <td>$S_0$</td>
        <td>$S_4$</td>
        <td>$S_1$</td>
        <td>$S_5$</td>
        <td>$S_2$</td>
        <td>$S_6$</td>
        <td>$S_3$</td>
        <td>$S_7$</td>
    </tr>
    <tr>
        <td>$S_0$</td>
        <td>$-(\lambda+\nu)$</td>
        <td>$\gamma$</td>
        <td>$\mu$</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
    </tr>
    <tr>
        <td>$S_4$</td>
        <td>$\nu$</td>
        <td>$-(\lambda+\gamma)$</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
    </tr>
    <tr>
        <td>$S_1$</td>
        <td>$\lambda$</td>
        <td>0</td>
        <td>$-(\lambda+\nu+\mu)$</td>
        <td>$\gamma$</td>
        <td>$\mu$</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
    </tr>
    <tr>
        <td>$S_5$</td>
        <td>0</td>
        <td>$\lambda$</td>
        <td>$\nu$</td>
        <td>$-(\lambda+\gamma)$</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
    </tr>
    <tr>
        <td>$S_2$</td>
        <td>0</td>
        <td>0</td>
        <td>$\lambda$</td>
        <td>0</td>
        <td>$-(\lambda+\nu+\mu)$</td>
        <td>$\gamma$</td>
        <td>$\mu$</td>
        <td>0</td>
    </tr>
    <tr>
        <td>$S_6$</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
        <td>$\lambda$</td>
        <td>$\nu$</td>
        <td>$-(\lambda+\gamma)$</td>
        <td>0</td>
        <td>0</td>
    </tr>
    <tr>
        <td>$S_3$</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
        <td>$\lambda$</td>
        <td>0</td>
        <td>$-(\mu+\nu)$</td>
        <td>$\gamma$</td>
    </tr> 
    <tr>
        <td>$S_7$</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
        <td>$\lambda$</td>
        <td>$\nu$</td>
        <td>$-\gamma$</td>
    </tr>
</table>

$$S_0:-(\lambda+\nu)P_0+\mu\cdot P_1+\gamma\cdot P_4=0$$
$$S_4:\nu\cdot P_0-(\lambda+\gamma)\cdot P_4=0$$

$$S_1:\lambda\cdot P_0-(\lambda+\mu+\nu)\cdot P_1+\mu\cdot P_2+\gamma\cdot P_5=0$$
$$S_5:\nu\cdot P_1+\lambda\cdot P_4-(\lambda+\gamma)\cdot P_5=0$$

$$S_2:\lambda\cdot P_1-(\lambda+\mu+\nu)\cdot P_2+\mu\cdot P_3+\gamma\cdot P_6=0$$
$$S_6:\nu\cdot P_2+\lambda\cdot P_5-(\gamma+\lambda)\cdot P_6=0$$

$$S_3:\lambda\cdot P_2-(\mu+\nu)\cdot P_3+\gamma\cdot P_7=0$$
$$S_7:\nu\cdot P_3+\lambda\cdot P_6-\gamma\cdot P_7=0$$

Из графа и таблицы видна несимметричность структуры и невыпонение условия эквивалентных преобразований.

Сложим уравнения:
$$-(\lambda+\nu)P_0+\mu\cdot P_1+\gamma\cdot P_4=0$$
$$\nu\cdot P_0-(\lambda+\gamma)\cdot P_4=0$$

$$-\lambda\cdot P_0+\mu\cdot P_1-\lambda\cdot P_4=0\Rightarrow \lambda\cdot P_0^v=\mu\cdot P_1$$

Сложим уравнения:
$$\lambda\cdot P_0-(\lambda+\mu+\nu)\cdot P_1+\mu\cdot P_2+\gamma\cdot P_5=0$$
$$\nu\cdot P_1+\lambda\cdot P_4-(\lambda+\gamma)\cdot P_5=0$$

$$\lambda\cdot(P_0+P_4)-\lambda\cdot(P_1+P_5)+\mu\cdot(P_2-P_1)=0\Rightarrow \lambda\cdot P_0^v-\lambda\cdot P_1^v+\mu\cdot(P_2-P_1)=0$$

Сложим уравнения:
$$\lambda\cdot P_1-(\lambda+\mu+\nu)\cdot P_2+\mu\cdot P_3+\gamma\cdot P_6=0$$
$$\nu\cdot P_2+\lambda\cdot P_5-(\gamma+\lambda)\cdot P_6=0$$

$$\lambda\cdot (P_1+P_5)-\lambda\cdot(P_2+P_6)+\mu\cdot(P_3-P_2)=0\Rightarrow \lambda\cdot P_1^v-\lambda\cdot P_2^v+\mu\cdot(P_3-P_2)=0$$

Сложим уравнения:
$$\lambda\cdot P_2-(\mu+\nu)\cdot P_3+\gamma\cdot P_7=0$$
$$\nu\cdot P_3+\lambda\cdot P_6-\gamma\cdot P_7=0$$
$$\lambda\cdot(P_2+P_6)-\mu\cdot P_3=0\Rightarrow \lambda\cdot P_2^v-\mu\cdot P_3=0$$

Т.е. окончательно получаем следующую систему уравнений:
$$\begin{cases}
  \lambda\cdot P_0^v=\mu\cdot P_1\\
  \lambda\cdot P_0^v-\lambda\cdot P_1^v=\mu\cdot(P_1-P_2)\\
  \lambda\cdot P_1^v-\lambda\cdot P_2^v=\mu\cdot(P_2-P_3)\\
  \lambda\cdot P_2^v=\mu\cdot P_3
\end{cases}$$

Получаем следующие условия эквивалентирования:
$$\frac{P_0^v}{P_1}=\frac{P_0^v-P_1^v}{P_1-P_2}=\frac{P_1^v-P_2^v}{P_2-P_3}=\frac{P_2^v}{P_3}=\frac{\mu}{\lambda}$$

Конечно, непосредственно их выполнить не удается, ведь нам неизвестны $P_1, P_2, P_3$.

Рассмотрим более простой случай, в котором все полностью симметрично:

<center><img src="dia8.png" width=300 hight=300></center>

$$-(\lambda+\nu)\cdot P_0+\mu\cdot P_1+\gamma\cdot P_2=0\tag{3.9}$$
$$-(\mu+\nu)\cdot P_1+\gamma\cdot P_3+\lambda\cdot P_0=0\tag{3.10}$$
$$-(\gamma+\lambda)\cdot P_2+\nu\cdot P_0+\mu\cdot P_3=0\tag{3.11}$$
$$-(\mu+\gamma)\cdot P_3+\lambda\cdot P_2+\nu\cdot P_1=0\tag{3.12}$$

Сложим уравнения (3.9), (3.11) или (3.10) и (3.12):

$$\lambda\cdot(P_0+P_2)=\mu\cdot(P_1+P_3)\Rightarrow \lambda\cdot P_0^v=\mu\cdot P_1^v$$

Если же мы уберем интенсивность $\mu$ между $S_2, S_3$, то (3.11) уравнение будет иметь вид:
$$-(\gamma+\lambda)\cdot P_2+\nu\cdot P_0=0$$

Тогда складывая (3.9) и (3.11), получим:
$$\lambda\cdot(P_0+P_2)=\mu\cdot P_1$$
Значит нужно снижать $\mu$ до некоторого значения $\mu_1$, чтобы 
$$\lambda\cdot P_0^v = \mu_1\cdot P_1^v$$

На практике для квазиэквивалентности берут $\mu_1=P_0^h\cdot \mu$ - снижается на долю времени, в которой система работоспособна.

В нашем случае $\mu_1=\frac{12}{17}\cdot 0.8\approx 0.56$

Тогда граф состояний будет выглядеть следующим образом:

<center><img src="ris_9.png" width=700 height=700></center>

$$P_0=\left(1+\frac{\lambda}{\mu_1}+\frac{\lambda^2}{\mu_1^2}+\frac{\lambda^3}{\mu_1^3}\right)^{-1}$$
$$P_1 = \frac{\lambda}{\mu_1}\cdot P_0$$
$$P_2 = \frac{\lambda^2}{\mu_1^2}\cdot P_0$$
$$P_3 = \frac{\lambda^3}{\mu_1^3}\cdot P_0$$

In [ ]:
[[P0 = 2725888/35071493,P1 = 4181760/35071493,P2 = 6767200/35071493,
         P3 = 11081500/35071493,P4 = 619520/35071493,P5 = 1232000/35071493,
         P6 = 2098000/35071493,P7 = 6365625/35071493]]

In [69]:
#Значения вероятностей для укрупненных состояний
lambda <- 1
mu1 <- 0.56

P0 <- (1+lambda/mu1+lambda^2/mu1^2+lambda^3/mu1^3)^-1
P1 <-P0*lambda/mu1
P2<-P0*lambda^2/mu1^2
P3<-P0*lambda^3/mu1^3
cat("P0 = ",P0,"\n")
cat("P1 = ",P1,"\n")
cat("P2 = ",P2,"\n")
cat("P3 = ",P3,"\n")

P0 =  0.08569912 
P1 =  0.1530341 
P2 =  0.2732752 
P3 =  0.4879915 


In [68]:
#Фактические значения вероятностей
cat("P0=",2725888/35071493+619520/35071493,"\n")
cat("P1=",4181760/35071493+1232000/35071493,"\n")
cat("P2=",6767200/35071493+2098000/35071493,"\n")
cat("P3=",11081500/35071493+6365625/35071493,"\n")

P0= 0.09538824 
P1= 0.1543635 
P2= 0.2527751 
P3= 0.4974731 


Из представленных расчетов видно, что даже квазиэквивалентное укрупнение состояний способно давать во многих случаях приемлемые результаты.

Попробуем методом укрупнения решить задачу 7 из лекции 5:

**Одноядерный микропроцессор ставит выполнямые задачи в очередь с ограничением не больше двух задач, иногда микропроцессор переключается на выполнение своих задач. Задача, которая выполняется в этот момент, помещается в очередь, в противном случае убивается. Интенсивность потока задач $\lambda$, потока выполнения задач $\mu$, интенсивность перечлючения операционной системы на выполнение своих задач $\nu$, интенсивность выполнения своих задач $\gamma$. Определить возможные состояния системы и финальные вероятности нахождения в них.** 

<center><img src="dia3_3.png" width=500 height=500></center>

Cистема уравнений для исходного графа имеет, как было показано ранее, следующий вид:

$$-(\lambda+\nu)P_0+\mu\cdot P_1+\gamma\cdot P_4=0\tag{3.13}$$
$$\lambda\cdot P_0-(\lambda+\mu+\nu)\cdot P_1+\mu\cdot P_2+\gamma\cdot P_5=0\tag{3.14}$$
$$\lambda\cdot P_1-(\lambda+\mu+\nu)\cdot P_2+\mu\cdot P_3+\gamma\cdot P_6=0\tag{3.15}$$
$$\lambda\cdot P_2-(\mu+\nu)\cdot P_3=0\tag{3.16}$$
$$\nu\cdot P_0-(\lambda+\gamma)\cdot P_4=0\tag{3.17}$$
$$\nu\cdot P_1+\lambda\cdot P_4-(\lambda+\gamma)\cdot P_5=0\tag{3.18}$$
$$\nu\cdot P_2+\nu\cdot P_3+\lambda\cdot P_5-\gamma\cdot P_6=0\tag{3.19}$$

Вместо детального решения данной задачи используем метод укрупнения состояний. Для этого попробуем выполнить горизонтальное объединение состояний $S_0, S_1, S_2$ в одно укрупненное состояние $W_0$, а также объединение состояний $S_4, S_5, S_6$ в одно укрупненное состояние $W_1$. В этом случае граф состояний получит следующий вид:

<center><img src="dia10.png" width=300 height=300></center>

Для укрупненного графа состояний система примет следующий вид:
$$-(\nu+\lambda)\cdot P_0^h+\gamma\cdot P_1^h+\mu\cdot P_3=0\tag{3.20}$$
$$-\gamma\cdot P_1^h+\nu\cdot P_0^h+\nu\cdot P_3=0\tag{3.21}$$
$$-(\mu+\nu)\cdot P_3+\lambda\cdot P_0^h=0\tag{3.22}$$

Таким образом, вместо 7 уравнений (6 линейно независимых) получили только 3 (2 линейно независимых) уравнения. Причем фактически укрупнение состояний сводится к тому, что мы вместо вероятностей нахождения в каждом из $S_0, S_1, S_2$ состояний определяем вероятность $P_0^h$ находиться или в $S_0$, или в $S_1$, или в $S_2$, которая конечно равна сумме вероятностей $P_0+P_1+P_2$. Аналогично, $P_1^h=P_4+P_5+P_6$. Таким образом, получаем уравнения связи с исходной задачей:

$$P_0^h=P_0+P_1+P_2\tag{3.23}$$
$$P_1^h=P_4+P_5+P_6\tag{3.24}$$

Заметим, что складывая уравнения (3.13)+(3.14)+(3.15)+(3.16) получим уравнение (3.21):

In [1]:
kill(lambda,nu,mu,gamma)$
expand(-(lambda+nu)*P0+mu*P1+gamma*P4+
      lambda*P0-(lambda+mu+nu)*P1+mu*P2+gamma*P5+
      lambda*P1-(lambda+mu+nu)*P2+mu*P3+gamma*P6+
      lambda*P2-(mu+nu)*P3);

(%o1)   P6 gamma + P5 gamma + P4 gamma - P3 nu - P2 nu - P1 nu - P0 nu

Или, складывая (3.17)+(3.18)+(3.19) тоже получаем (3.21):

In [2]:
expand(nu*P0-(lambda+gamma)*P4+
      nu*P1+lambda*P4-(lambda+gamma)*P5+
      nu*P2+nu*P3+lambda*P5-gamma*P6);

(%o2)  - P6 gamma - P5 gamma - P4 gamma + P3 nu + P2 nu + P1 nu + P0 nu

Если сложить (3.20)+(3.22), то тоже получаем (3.21).

Однако система (3.13)-(3.19) не будет полностью эквивалентна системе (3.20)-(3.22), поскольку сопоставляя уравнения (3.22) и (3.16) видно, что $P_0^h$ фактически совпадает с  $P_2$, чего по факту быть не может.

Это подтверждают и расчеты.

Расчеты вероятностей $P_0,P_1,\ldots, P_6$ были выполнены ранее:

In [3]:
lambda:1$ nu:0.5$ mu:0.8$ gamma:1.2$
solve([-(lambda+nu)*P0+mu*P1+gamma*P4=0,
      lambda*P0-(lambda+mu+nu)*P1+mu*P2+gamma*P5=0,
      lambda*P1-(lambda+mu+nu)*P2+mu*P3+gamma*P6=0,
      lambda*P2-(mu+nu)*P3=0,
      nu*P0-(lambda+gamma)*P4=0,
      nu*P1+lambda*P4-(lambda+gamma)*P5=0,
      P0+P1+P2+P3+P4+P5+P6=1],
      [P0,P1,P2,P3,P4,P5,P6]);


rat: replaced -1.5 by -3/2 = -1.5

rat: replaced 0.8 by 4/5 = 0.8

rat: replaced 1.2 by 6/5 = 1.2

rat: replaced -2.3 by -23/10 = -2.3

rat: replaced 0.8 by 4/5 = 0.8

rat: replaced 1.2 by 6/5 = 1.2

rat: replaced -2.3 by -23/10 = -2.3

rat: replaced 0.8 by 4/5 = 0.8

rat: replaced 1.2 by 6/5 = 1.2

rat: replaced -1.3 by -13/10 = -1.3

rat: replaced 0.5 by 1/2 = 0.5

rat: replaced -2.2 by -11/5 = -2.2

rat: replaced 0.5 by 1/2 = 0.5

rat: replaced -2.2 by -11/5 = -2.2


             402688        617760        999700        769000         91520
(%o7) [[P0 = ───────, P1 = ───────, P2 = ───────, P3 = ───────, P4 = ───────, 
             3951293       3951293       3951293       3951293       3951293
                                                        182000        888625
                                                   P5 = ───────, P6 = ───────]]
                                                        3951293       3951293

In [4]:
float(%);

(%o8) [[P0 = 0.10191296874213074, P1 = 0.15634375886576876, 
P2 = 0.2530057882318522, P3 = 0.19461983710142478, P4 = 0.02316203835048426, 
P5 = 0.046060871719713015, P6 = 0.22489473698862625]]

In [5]:
/* P0h */
0.1019129687421307+0.1563437588657688+0.2530057882318522;
/* P1h */
0.02316203835048426+0.04606087171971301+0.2248947369886262;

(%o9)                         0.5112625158397517

(%o10)                        0.2941176470588235

In [6]:
solve([-(lambda+nu)*P0h+gamma*P1h+mu*P3=0, 
       -gamma*P1h+nu*P0h+nu*P3=0,
       P0h+P1h+P3=1],
       [P0h,P1h,P3]);


rat: replaced -1.5 by -3/2 = -1.5

rat: replaced 1.2 by 6/5 = 1.2

rat: replaced 0.8 by 4/5 = 0.8

rat: replaced 0.5 by 1/2 = 0.5

rat: replaced -1.2 by -6/5 = -1.2

rat: replaced 0.5 by 1/2 = 0.5


                               156        5        120
(%o11)                 [[P0h = ───, P1h = ──, P3 = ───]]
                               391        17       391

In [7]:
float(%);

(%o12) [[P0h = 0.3989769820971867, P1h = 0.29411764705882354, 
                                                      P3 = 0.3069053708439898]]

Из полученных результатов видно, что правильно расчитано только $P_1^h$. Сопоставляя уравнения, противоречащие полной эквивалентности: 
$$\lambda\cdot P_2-(\mu+\nu)\cdot P_3=0$$
$$-(\mu+\nu)\cdot P_3+\lambda\cdot P_0^h=0\Rightarrow \lambda\cdot (P_0+P_1+P_2)-(\mu+\nu)\cdot P_3=0 $$

можно увидеть, что для квазиэквивалентирования интенсивность $\lambda$ нужно уменьшать на долю времени работоспособности системы в состоянии $W_0$, т.е. на $P_0+P_1+P_2$. Значит вместо уравнения $\lambda\cdot (P_0+P_1+P_2)-(\mu+\nu)\cdot P_3=0$ нужно взять $\lambda\cdot (P_0+P_1+P_2)^2-(\mu+\nu)\cdot P_3=0$. Или можно решить систему уравнений:

$$\begin{cases}
-(\nu+\lambda\cdot P_0^h)\cdot P_0^h+\gamma\cdot P_1^h+\mu\cdot P_3=0\\
-\gamma\cdot P_1^h+\nu\cdot P_0^h+\nu\cdot P_3=0
\end{cases}$$

In [8]:
solve([-(lambda*P0h+nu)*P0h+gamma*P1h+mu*P3=0, 
       -gamma*P1h+nu*P0h+nu*P3=0,
       P0h+P1h+P3=1],
       [P0h,P1h,P3]);


rat: replaced -0.5 by -1/2 = -0.5

rat: replaced 1.2 by 6/5 = 1.2

rat: replaced 0.8 by 4/5 = 0.8

rat: replaced 0.5 by 1/2 = 0.5

rat: replaced -1.2 by -6/5 = -1.2

rat: replaced 0.5 by 1/2 = 0.5


               sqrt(154921) - 221        5          sqrt(154921) - 461
(%o13) [[P0h = ──────────────────, P1h = ──, P3 = - ──────────────────], 
                      340                17                340
                        sqrt(154921) + 221        5        sqrt(154921) + 461
               [P0h = - ──────────────────, P1h = ──, P3 = ──────────────────]]
                               340                17              340

In [9]:
float(%);

(%o14) [[P0h = 0.5076472082735436, P1h = 0.29411764705882354, 
P3 = 0.19823514466763287], [P0h = - 1.8076472082735435, 
P1h = 0.29411764705882354, P3 = 2.51352956121472]]

По смыслу подходит первое множество решений. При этом видим довольно высокую точность:

<table>
    <tr>
        <td>Метод расчета</td>
        <td>
            $P_0^h$
        </td>
        <td>
            $P_1^h$
        </td>
        <td>$P_3$</td>
    </tr>
    <tr>
        <td>Точный метод</td>
        <td>0.5112625158397517</td>
        <td>0.2941176470588235</td>
        <td>0.19461983710142478</td>
    </tr>
    <tr>
        <td>Метод укрупнения</td>
        <td>0.5076472082735436</td>
        <td>0.29411764705882354</td>
        <td>0.19823514466763287</td>
    </tr>
</table>